### ISSUE TITLE
[XMLExchange] Folder Structure not created on Import

### ISSUE THREAD

**Issue by MilennialZero** [archimatetool/archi#288]
A different question (issue?):

In an effort to try to purge the ghost views, I deleted the ghost views, and exported the model to Open Exchange File. This actually works (I think) - the ghost views are no longer there - but the folders I've defined to organise my views don't get exported. Is there something I can do when exporting that will preserve the folders I've defined in the model for my views?
:-( 
![folders not there](https://user-images.githubusercontent.com/30424386/31316015-051450fe-ac70-11e7-8fe3-a02dca8d9469.png)
![export folder structure](https://user-images.githubusercontent.com/30424386/31316014-05124606-ac70-11e7-9962-5910d103f1aa.png)

**Comment by Phillipus** [comment]
Folders are exported in the Exchange format but currently, not imported. There are reasons for this -> how should Archi map a different folder structure from another tool onto its own folder structure.

Again, as with any feature request, should anyone wish to submit a patch to implement this, feel free.

**Comment by MilennialZero** [comment]
Yep. Understand. I suppose a simple "Whatever folders and whatever views in those folders in whatever sequence they appear in the source, import them just like that" is naïve? There's probably some subtlety in there that I can't see. You don't want me patching anything. Thanks again for your work.

**Comment by Phillipus** [comment]
Hi Jon, as you know Archi has a fixed folder pattern in the tree (Business, Application, Technology, etc) and fixed rules for where the elements go (Business elements in Business folder or sub-folder, relations in "Relations" folder). It's easy enough to export that structure in the exchange format but a tool that imports it might not actually support that structure itself. Exporting and importing from Archi will of course use the same folder structure, but as the exchange format is meant to be used from different tools, it becomes a problem if, say, the exporting tool allows Application elements mixed with Business elements. Where to put them?

So, a general rule was made - when importing put each element in its default folder. Now, I did play around with this when importing from another tool, but there were so many transpositions that had to be made that it effectively nullified the folder structure of the import folder.

**Comment by samperd** [comment]
Hey Folks,

I was hoping to use the export/import mechanism under a few different use cases, including:

1. creat a generic structure of an architecture including folders and then be able to re-use that structure through importing
2. Share this structure with an architecture team so all projects can start with the same strucutre
3. exchange models across groups (eg throughout a government).

I understand the issue with not wanting to import foreign structures automatically, however I have a few thoughts.

## Potential Implementations
1. when importing, present the user with a dialogue presenting check boxes for each of the elements that could be imported, eg:
* Objects
* Relationships
* Folders
* Views
* Add to existing model
  * Override NAME 
  * Override Purpose

2. When exporting, could we capture the software and version that has exported it (might be a request for TOGAF, could not find a place to provide feedback). Then if the file was exported from Archi and has folders then you get to keep those files

**Comment by Glomal** [comment]
Similar to Samperd I was hoping to use the export/import mechanism to create and maintain models with thousands of elements (meaning offline lists and folder structures are much more important than the modelling GUI.

So a few observations / comments / suggestions.

1) The XML import should operate more like the CSV import(s). If the model name (and ideally it would be a separate identifier, not just a name) is the same it should merge the XML data into the existing model - and only if the model name is different/unrecognised create a new model (the current behaviour).

2) Similarly with the folder structure - the import process should check for the existence of the folder specified in the XML data - and if present merge and if not create anew. The if element identifiers are defined in the XML to be "in" a folder they should be added to the folder (or overwritten if they already exist) - and if they are not defined in the XML in any folder then default to the "standard" folders. This I think would be aided if the folders had a separate id namespace from the model elements and relationships - and the folder name was just an attribute. Using this sort of mechanism it should be possible to rename the folders - even the "standard" ones. [PS It may not be the archimate standard but you should definitely have separate folders for Information/Data and Applications.

3) You should be able to have other "folder id namespaces" from the standard one to create different tree structures to which the lements and relationships "belong" and through which they may be viewed. This would also facilitate the creation of multidimensional matrices and tensors used to express other relationships than are simply visible through the 2D modelling tool (GUI).

4) Lastly, visual groups clould be aligned with folders in the Views structure. and maybe there is a special folder structure (different from the standard ones) that should align with element and relationship groupings.

5) There ought to be, not exactly a 'programmatic', but a more automated methods of generating some standard views based on inclusion in a folder or grouping - possibly defined using element properties? [ie maybe grouping membership should generate an automatic property of being a member of that grouping. Perhaps some sort of 'stereotyping' mechanism?]

**Comment by Glomal** [comment]
One last suggestion: there should be hooks into an XSLT processor during the import process such that you can feed in arbirtary model-containing XML file and a matching XSLT specification to translate it into an Archi Open Exchange XML file and then do the actual import into Archi. [And it should preserve folder structures - or translate them in accordance with the XSLT script.

**Comment by Phillipus** [comment]
There's lots of work involved in implementing those suggestions and would require a total re-write of Archi and other plug-ins.

For now, the focus of this issue is maintaining, if possible, the folder structure when importing or at least a best shot at it.

To quote my comments above:

> How should Archi map a different folder structure from another tool onto its own folder structure?

A quick example - a Business element is in the same folder as a Strategy element. If, when importing, Archi separates these into its folder structure doesn't that defeat the object of the original folder structure?

> Perhaps the real solution is to not have the rigid folder structure in Archi. 

Again, this would require a total re-write of Archi (and any downstream plug-ins).

**Comment by samperd** [comment]
@Phillipus , I think I was taking a pretty narrow view of "folders" initiatially. Now that I am back from vacation and looking at this problem with a fresh perspective I more than one "folder scenario" at play. Perhaps by describing these scenarios we can better understand what users (like me) want and what developers can work with.

First off I see three levels of folder scenarios.

1. The first layer are the parent folders defined by archi as part of the whole model where elements live according to their domain. (Strategy, Business, Application, including Views). These folders are the default archi folders and created for each new project.  (level 1)
3. The Second level is found as sub folders under each domain (eg Strategy/Dir1). These sub folders are created by the user through the use of the right-click context menu option "New > Folder" (level 1.A)
2. The Third level is found within the folder for the views, which some users often create new directories using the context menu as in scenario 2. These folders seem to be where most heavy customization occurs (Level 1.B)

Those are the three levels I see.

As primarily an Archi user, my goal would be to harmonize the folder structure of a foreign model to the Archi model.

## Level 1
**Assumptions**:
* There must be a number of assumptions and linkages to this folder structure in the logic of the software, but lets stick with the user view for now.
* The default folders should be the minimum requirement.
* The question then becomes what to do with folders that fit outside of this Archi default structure.

**Proposal**
* When importing models, if the default folders do not exist, then create them.
* If folders exist outside of this structure then create sub-folders according to Proposals for Level 1.A and Level 1.B
* All domain elements will exist in sub-folders under their domain
* All relationships will exist in sub-folders under Relations
* All Views will exist in sub-folders under "Views" 


## Level 1.A
**Assumptions**:
* users can create new and arbitrary sub-folders
* These sub-folders will contain only elements of the parent domain. For instance a sub-folder under "Strategy" can in turn only contain "Strategy" Elements.
* All relationships between elements will be captured in parent "Relations" folder

**Proposal:**
* If an imported model has a "Strategy/ImportedDir" directory that contains both "Strategy" and "Business" elements then I propose that the following directories are created in Archi "Strategy/ImportedDir" and "Business/ImportedDir"

This way all folders and elements are imported, but their location might move, while maintaining the semantics of the sub-folders.

## Level 1.B
Assumptions:
* users can create new and arbitrary sub-folders
* All views will exist either under the Parent View folder or within a sub-folder of Views

Proposal
* if an imported model contains views in subfolders under View parent, then import folders as is.
* If an imported model contains views outside of the View parent folder, consolidate these views under Views Folder
* If an imported model contains views in sub-folders outside of Views Parent folder, then create a sub-folder for those views and import only the views portion of those folders. Domain elements would be imported to sub-folders under each domain.

## Other Notes:
In order to maintain some governance of models and elements as they get passed around it would be good to ensure we have some ability to trace lineage of model elements.

**Exporting Models From Archi**
* Elements exported by Archi should all contain the following properties (unless a GUI menu is selected to no include these at some point):
    * Source Model: The complete file path of the model
    * Archi Version: What version of Archi exported the model
    * Export Date: 

**Importing Models To Archi**
* Elements imported by Archi should all contain the following properties (unless a GUI menu is selected to no include these at some point):
    * Import existing Properties
    * Originating Folder: To document when the element came from in the event that Archi changes the folder structure.
    * perhaps we need an element that can capture the metadata of the imported model and associated elements? (eg Note Element existing in an "Import" View


I hope these additional comments help to clarify what I am looking for at least. Perhaps others have different views or goals.

Cheers

**Comment by Phillipus** [comment]
Thanks for the feedback and I shall chew it over these ideas, some of which I can see are good ideas.

But for now, a quick comment in response to this:

> In order to maintain some governance of models and elements as they get passed around it would be good to ensure we have some ability to trace lineage of model elements.

Once a model is exported in the exchange format there is absolutely no guarantee what will be preserved or changed after that. So these properties might be gone. The exchange format is not suited to round-tripping.

**Comment by Phillipus** [comment]
More..

I think there you have presented some good ideas to implement at least something better than what we have now (nothing).

**Comment by Phillipus** [comment]
> **Exporting Models From Archi**
> * Elements exported by Archi should all contain the following properties (unless a GUI menu is selected to no include these at some point):
    * Source Model: The complete file path of the model
    * Archi Version: What version of Archi exported the model
    * Export Date: 

It's possible to export additional information in properties to the exported file. However, be aware that another importing tool might discard that information. There are no guarantees. Thus, when Archi re-imports a model it cannot expect to see those properties again. Many times it's a one way trip, from Archi to Tool X, or from Tool X to Archi.

If this information was (re-)imported into Archi these would become Archi properties and, if the model was re-exported, a check would need to be made to export these ones and they may require updating. Do you update the properties that are now in the Archi model and re-export, or do you leave those alone and re-export new updated properties?

> **Importing Models To Archi**
> * Elements imported by Archi should all contain the following properties (unless a GUI menu is selected to no include these at some point):
    * Import existing Properties
    * Originating Folder: To document when the element came from in the event that Archi changes the folder structure.
    * perhaps we need an element that can capture the metadata of the imported model and associated elements? (eg Note Element existing in an "Import" View

Possibly.

But now I am concerned that we are straying into the "tail wagging the dog" syndrome where we have to be careful that the ArchiMate Exchange Format does not dictate how tools are written or force them to add features. Are other tools vendors re-writing their tools to accommodate Archi?

So, I might suggest a starting point based on your ideas. To implement this is not trivial and would take some time and, as there is only one of me, it's not at the top of my to do list.

What would be useful is a more fleshed out algorithm for doing at least a better job of slotting imported concepts and folders into Archi's sub-folders. Such as:

- organisation item is a folder
 -- iterate through children of folder
 -- are children of different types?
 -- create sub-folders as necessary. store in lookup table
- organisation item is a concept
 -- put this in a default folder
- organisation item is a view
-- etc...

**Comment by Phillipus** [comment]
For reference, this is the source code for importing:

https://github.com/archimatetool/archi/blob/master/org.opengroup.archimate.xmlexchange/src/org/opengroup/archimate/xmlexchange/XMLModelImporter.java

lines 121 and 362-393

**Comment by samperd** [comment]
@Phillipus 

> But now I am concerned that we are straying into the "tail wagging the dog" syndrome where we have to be careful that the ArchiMate Exchange Format does not dictate how tools are written or force them to add features. Are other tools vendors re-writing their tools to accommodate Archi?

I am not sure what is happening with other tool vendors, however the sense I get from The Open Group is that Archi is the reference implementation for import and export of the Archimate Exchange Format. THis becomes a double edged sword however. I feel on one hand there is an opportunity to influence how the Exchange format is used, while at the same time there is a risk that as Archi changes how it treats import/export it might impact the reference implementation other vendors may mimick.

>  Q. Is there a source code implementation available?
> 
> A. A plugin has been developed for the Archi tool and is distributed with the binary release. Source code is available in the GitHub repository at https://github.com/archimatetool/OpenGroupXMLExchange .

source: https://www.opengroup.org/open-group-archimate-model-exchange-file-format

Not sure if there is a project governance issue here or not, but I am coming at this from the perspective of a user. I certainly don't want to break conformance to the standard, but I am willing to explain proposed changes to the standard if required.

**Comment by Phillipus** [comment]
The "organization" structure in the exchange format is optional. It was meant to optionally represent a hierarchical organization that might or might not exist in a tool. I could imagine an ArchiMate service (a "tool") that did not have such a structure.

The issue for Archi is simply this - something better than what we have now when importing has not been implemented, and it has not been implemented because I haven't had the time to do it. So for me it's not about governance it's simply that I have to code everything. If only this project was open source... ;-)

**Comment by samperd** [comment]
@Phillipus 

> It's possible to export additional information in properties to the exported file. However, be aware that another importing tool might discard that information. There are no guarantees. Thus, when Archi re-imports a model it cannot expect to see those properties again. Many times it's a one way trip, from Archi to Tool X, or from Tool X to Archi.

I have not looked into the exchange format enough to advise on specific changes, however am I to understand correctly that "Properties" is specific to Archi and there is no equivalent in the exchange format?

If this is archi specific, would adding an Archi namespace for "property" elements as an extension to the standard be an appropriate way to handle this in the short term? Then, if it meets our needs we can propose the extension change to be included in the core standard?

in my working environment (Canadian Federal Government) we have a few vendors playing in the EA space which I could try to find willing testers to ensure we have interoperability. The key to interoperability in the Canadian Government will come from Our directive on IT management

> Application Architecture
> 
>     C.2.3.8Use Open Standards and Solutions by Default
>        C.2.3.8.1 Where possible, use open standards and open source software first
>        C.2.3.8.2 If an open source option is not available or does not meet user needs, favour platform-agnostic COTS over proprietary COTS, avoiding technology dependency, allowing for substitutability and interoperability

source: https://www.tbs-sct.gc.ca/pol/doc-eng.aspx?id=15249

Key EA Vendor Solutions I know of:
* Archi
* Sparx EA
* Qualiware

Maybe a paralel step in this process is to examine what happens for a round trip scenario using the archimate exchange samples and the above listed vendors. If this is desired then i can spin off a side effort as a separate issue.

**Comment by Phillipus** [comment]
> however am I to understand correctly that "Properties" is specific to Archi and there is no equivalent in the exchange format?

No. Properties are in the exchange format and I think most, if not all, tools support them.

**Comment by samperd** [comment]
@Phillipus 

> What would be useful is a more fleshed out algorithm for doing at least a better job of slotting imported concepts and folders into Archi's sub-folders. Such as:
>
> organisation item is a folder
> -- iterate through children of folder
> -- are children of different types?
> -- create sub-folders as necessary. store in lookup table
> organisation item is a concept
> -- put this in a default folder
> organisation item is a view

I see how a more defined algorithm will make things easier, however from a user perspective I am not sure how much light I can shed on this other than the desired end state.

Now that I have actually taken a look at the [actual standard from OG](https://publications.opengroup.org/c174) I see that our first issue is that the "Organization" container as described in chapter 10 is not tightly defined. This will make coming up with an algorithm tough as we are further specifying expected behaviour of the expressed standard. (I usually advocate strongly for standards).

> 10 Organizational Information
>
>This chapter describes an optional container for providing information about the organization of model elements and relations within the exchange file.
>Model elements and relations may be referenced by means of identifiers to represent logical relationships among the units of content in the exchange file. More than one logical organization may be described.
>
>10.1 Organization
>
>An optional container for organizing units of content in the model using item elements and item references.
>An organization has no meaning unless it has at least one item element.

What about working through a couple of initial iterations of improving this approach. For instance:

Iteration 1:
* Start a new "plugin" as to not drastically impact the core functionality of archi until we have something a bit more solid.
* if the folders fit nicely into Archi's structure, import as is. (This should satisfy exchanges between Archi instances)
* If the folder structure differs from Archi then import full model to NEW folder called "etc" or "src" and build the complete model under that node. Any new elements will remain in Archi's structure. If Archi can't handle this structure, then allow the user to manualy click and drag required elements to the appropriate Archi folders. A painful hack perhaps

Iteration 2:
* for models imported to "etc", automatically copy or reference those elements into the appropriate folders for archi.

Iteration 3:
* figure out a means for multiple folder models to co-exist and not collide.
* Alternatively propose changes to standard for folder structure standards to be more well defined, hopefully falling close to Archi's existing model.


I am not sure where to go from here.

P.S If Archi was written in python or used XSLT I could help out with prototyping, but I never made friends with Java

**Comment by Phillipus** [comment]
> Alternatively propose changes to standard for folder structure standards to be more well defined, hopefully falling close to Archi's existing model.

I can't see how this can be more well defined, given the fact that it is not designed specifically to represent folders (although that is how it is used), especially not Archi's. It is an optional organization of objects that might be, for example, a table of contents. There are other tools and services that might differ completely.

**Comment by Phillipus** [comment]
I now have a solution. It is a simple and elegant solution and it's better than what we have now. I am not going to add additional Properties or other stuff.

#### If the imported XML file has come from Archi (and not tampered with), the folder structure will be preserved.

Even if we did no more than this it will satisfy a lot of users. :-)

#### If the imported XML file has a foreign organization/item structure then this happens:

1. If the item references a concept or view the item is put into its default Archi folder ("Business", "Application", etc)
1. If the item is a child in an item hierarchy (folders) then recreate that folder hierarchy under the default folder for that concept/view.
1. Move the item to the sub-folder.

This process will separate each object into its matching Archi folder or sub-folder. Folder hierarchies will be preserved under each Archi folder.

If an item contains a mixture of different object types then the folder structure will be created for each object type in its folder. For example in this pseudo-XML:

```
<item Folder1>
   <item Folder2>
      <item BusinessActor>
      <item ApplicationComponent>
   </item>
</item>
```

Will appear in Archi as:

```
Business
    |----Folder1
             |---- Folder2
                       |---- BusinessActor
Application
    |----Folder1
             |---- Folder2
                       |---- ApplicationComponent
```

It needs more testing, particularly with edge cases and foreign item structures.

**Comment by samperd** [comment]
@Phillipus 

> If the imported XML file has come from Archi (and not tampered with), the folder structure will be preserved.

+1

> If the imported XML file has a foreign organization/item structure then this happens:
> 
> * If the item references a concept or view the item is put into its default Archi folder ("Business", "Application", etc)
> * If the item is a child in an item hierarchy (folders) then recreate that folder hierarchy under the default folder for that concept/view.
> * Move the item to the sub-folder.

+1

> If an item contains a mixture of different object types then the folder structure will be created for each object type in its folder.

+1 


This proposal is in line with the same line of thinking proposed above. Nice articulation.

I am looking forward to beta testing.

**Comment by Phillipus** [comment]
Forgot to say - this follows @samperd's ideas in his comments above:

> If an imported model has a "Strategy/ImportedDir" directory that contains both "Strategy" and "Business" elements then I propose that the following directories are created in Archi "Strategy/ImportedDir" and "Business/ImportedDir"

> This way all folders and elements are imported, but their location might move, while maintaining the semantics of the sub-folders.

**Comment by Phillipus** [comment]
> This proposal is in line with the same line of thinking proposed above. Nice articulation.

Yes, noted in previous comment (missed out from my main comment)

**Comment by Phillipus** [comment]
See https://github.com/archimatetool/archi/blob/78130d032a5040ef59f51864ba4b036e9151bd1c/org.opengroup.archimate.xmlexchange/src/org/opengroup/archimate/xmlexchange/XMLModelImporter.java#L367

**Comment by Phillipus** [comment]
> I am looking forward to beta testing.

I'm aiming for Archi 4.6 public beta in a few weeks.

For those brave souls who have set up Eclipse with Archi's code, it can be tested there. :-)

**Comment by Phillipus** [comment]
One wrinkle - empty folders/items.

If a folder/item is empty we have no way of knowing where to put it unless it has come from Archi and its top level folder/item matches one of Archi's top level folders.

I'm working on that refinement now.

**Comment by Phillipus** [comment]
Another thing - import the documentation for each folder.

Note - this version of the exchange format does not support Properties on folders.

**Comment by Phillipus** [comment]
Should mostly be done now.

**Comment by Phillipus** [comment]
Committed to master branch.

**Comment by Phillipus** [comment]
Fixed in 4.6
